Add new WorkerPoolHash calculation to rotate nodes on OSCKey change

MichaelEischer opened the pull request:
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it:**

**Based on #9846, please review that PR first**

This PR implements version 2 of the gardener node agent secret name calculation, as proposed in #9699. Version 2 of the name calculation moves most fields that were included in the `WorkerPoolHash` calculation to the name calculation of the `OperatingSystemConfig`. These fields are removed from the `WorkerPoolHash` calculation and are replaced with a field `OSCHash` that includes the just mentioned secret name. This allows Gardener to trigger a node pool roll if there are relevant changes in the `OperatingSystemConfig`. In addition, it is possible in the future to include further node roll triggers without requiring changes to the provider extensions.

The secret name uses `gardener-node-agent-hash(worker.CRI, machineType, volume type+size, worker.Machine.Image.Name+Version, minorKubernetesVersion, credentialsRotationStatus, nodeLocalDNS, kubeReserved, evictionHard.Memory, cpuManagerPolicy)[:16]-<worker.Name>-<suffix>`. For `kubeReserved` currently all fields are included. For `evictionHard`, I've so far only included memory, although we could also add the FS-based limits.

The PR also includes a new feature gate `WorkerPoolHashV2`. This is necessary, as the provider extensions first have to vendor the new Gardener code, before the new hash calculation for the OperatingSystemConfig can be used. Thus, the `OSCHash` field must remain unused until the provider extensions are able to interpret it to prevent node rolls when upgrading the extension.

**Which issue(s) this PR fixes:**
Part of #9699
Fixes https://github.com/gardener/gardener/issues/9265 (see https://github.com/gardener/gardener/issues/9265#issuecomment-2099907212 regarding the `cpuManagerPolicy` field)

**Special notes for your reviewer:**

**Based on #9846, please review that PR first**

To test the changes locally, enable the feature gate first. Then reconcile the shoot once to ensure that the `pool-hashes` secret (from #9846) is up to date.

```diff
diff --git a/example/gardener-local/gardenlet/values.yaml b/example/gardener-local/gardenlet/values.yaml
index 063ef2b12..ab2bd1288 100644
--- a/example/gardener-local/gardenlet/values.yaml
+++ b/example/gardener-local/gardenlet/values.yaml
@@ -45,6 +45,7 @@ config:
     # Seed config. Only when this is set, gardenlet's behavior changes.
     IPv6SingleStack: true
     ShootManagedIssuer: true
+    NewWorkerPoolHash: true
   etcdConfig:
     featureGates:
       UseEtcdWrapper: true
```

- [ ] I think that the changed hash calculation and switching between the versions should be E2E tested. However, I'm unsure how to proceed here. Create a new E2E test? Extend an existing one? In particular, the feature gate would have to be enabled, which probably isn't the case right now for the E2E tests?

**Release note**:
<!--
Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|noteworthy|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```breaking developer
The function `github.com/gardener/gardener/extensions/pkg/controller/worker/WorkerPoolHash` now expects separate additional data for version 1 and 2 of the hash calculation. Version 2 does not include extension provider specific fields by default. Add those to the additional data for version 2 if necessary.
```
```feature operator
Version 2 of the WorkerPoolHash calculation now takes the current name of the OperatingSystemConfig into account. Its usage is controlled by the gardenlet feature gate `NewWorkerPoolHash`. All provider extension must be upgraded before enabling this feature gate. The new calculation also ensures that changes of the fields kubeReserved, evictionHard and cpuManagerPolicy in the kubelet config of a worker pool result in node rolls.
```

MichaelEischer committed:
OSC: implement V2 secret name generation

MichaelEischer committed:
worker: introduce WorkerPoolHashV2 based on OSCKey

MichaelEischer committed:
OSC: move worker pool hash enablement behind feature flag

MichaelEischer committed:
OSC: Minimal test fix

MichaelEischer committed:
OSC: port tests from WorkerPoolHash for KeyV2

MichaelEischer committed:
OSC: also consider CPUManagerPolicy in KeyV2

MichaelEischer committed:
OSC: consider EvictionHard for KeyV2

MichaelEischer committed:
rename oscHash to nodeAgentSecretName

rfranzke commented on pkg/apis/core/v1beta1/helper/helper.go:
```diff
@@ -1162,6 +1162,19 @@ func CalculateEffectiveKubernetesVersion(controlPlaneVersion *semver.Version, wo
 
+// CalcluateEffectiveResourceReservations if a shoot has a kubelet configuration specified at the worker group, return this,
+// otherwise the shoot resource reservations
+func CalcluateEffectiveResourceReservations(shootKubelet *gardencorev1beta1.KubeletConfig, workerKubernetes *gardencorev1beta1.WorkerKubernetes) *gardencorev1beta1.KubeletConfigReserved {
```
add a unit test

MichaelEischer replied:
Done

rfranzke commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -921,17 +929,20 @@ func (o *operatingSystemConfig) calculateKeyForVersion(version int, worker *gard
+	kubeReserved := v1beta1helper.CalcluateEffectiveResourceReservations(o.values.KubeletConfig, worker.Kubernetes)
 
-	return CalculateKeyForVersion(version, kubernetesVersion, worker)
+	return CalculateKeyForVersion(version, kubernetesVersion, o.values.CredentialsRotationStatus, worker, o.values.NodeLocalDNSEnabled, kubeReserved)
```
Nit: Why not passing `o.values` to make it easier to extend this in the future?

MichaelEischer replied:
Is changed.

rfranzke commented on docs/deployment/feature_gates.md:
```diff
@@ -34,6 +34,7 @@ The following tables are a summary of the feature gates that you can set on diff
 | ShootManagedIssuer                 | `false` | `Alpha` | `1.93` |        |
 | VPAForETCD                         | `false` | `Alpha` | `1.94` |        |
 | VPAAndHPAForAPIServer              | `false` | `Alpha` | `1.95` |        |
+| WorkerPoolHashWithNodeAgentSecret  | `false` | `Alpha` | `1.96` |        |
```
1.97

rfranzke commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -921,17 +929,20 @@ func (o *operatingSystemConfig) calculateKeyForVersion(version int, worker *gard
+func calculateKeyForVersion(oscVersion int, kubernetesVersion *semver.Version, credentialsRotation *gardencorev1beta1.ShootCredentialsRotation, worker *gardencorev1beta1.Worker, nodeLocalDNSEnabled bool, kubeReserved *gardencorev1beta1.KubeletConfigReserved) (string, error) {
 	switch oscVersion {
 	case 1:
 		return KeyV1(worker.Name, kubernetesVersion, worker.CRI), nil
```
Can you add a  TODO statement to remove this `KeyV1` when support for Kubernetes `1.30` is dropped?

MichaelEischer replied:
I've added a todo.

rfranzke commented on docs/deployment/feature_gates.md:
```diff
@@ -206,3 +207,4 @@ A *General Availability* (GA) feature is also referred to as a *stable* feature.
 | ShootManagedIssuer              | `gardenlet`                       | Enables the shoot managed issuer functionality described in GEP 24.                                                                                                                                                                                                                                                                     ...
 | VPAForETCD                      | `gardenlet`, `gardener-operator`  | Enables VPA for `etcd-main` and `etcd-events`, regardless of HVPA enablement.                                                                                                                                                                                                                                                           ...
 | VPAAndHPAForAPIServer           | `gardenlet`                       | Enables an autoscaling mechanism for shoot kube-apiserver where it is scaled simultaneously by VPA and HPA on the same metric (CPU and memory usage). The pod-trashing cycle between VPA and HPA scaling on the same metric is avoided by configuring the HPA to scale on average usage (not on average utilization) and by picking the ...
+| WorkerPoolHashWithNodeAgentSecret | `gardenlet`                       | Enables usage of the new worker pool hash calculation. All provider extension must be upgraded to support this feature first.enablement.                                                                                                                                                                                              ...
```
typo in description?

rfranzke commented on pkg/features/features.go:
```diff
@@ -81,7 +81,13 @@ const (
+	// WorkerPoolHashWithNodeAgentSecret enables a new calculation method for the worker pool hash used by provider extensions. All
+	// such extensions must be upgraded to this Gardener version before enabling the feature gate.
+	// owner: @MichaelEischer
+	// alpha: v1.96.0
```
97

rfranzke commented on extensions/pkg/controller/worker/machines.go:
```diff
@@ -93,7 +93,15 @@ func (m MachineDeployments) HasSecret(secretName string) bool {
 
 // WorkerPoolHash returns a hash value for a given worker pool and a given cluster resource.
-func WorkerPoolHash(pool extensionsv1alpha1.WorkerPool, cluster *extensionscontroller.Cluster, additionalData ...string) (string, error) {
+func WorkerPoolHash(pool extensionsv1alpha1.WorkerPool, cluster *extensionscontroller.Cluster, additionalDataV1 []string, additionalDataV2 []string) (string, error) {
```
pls adapt https://github.com/gardener/gardener/blob/master/docs/extensions/worker.md

rfranzke commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -947,7 +947,14 @@ func (o *operatingSystemConfig) calculateKeyForVersion(version int, worker *gard
+	worker *gardencorev1beta1.Worker,
+	nodeLocalDNSEnabled bool,
+	kubeReserved *gardencorev1beta1.KubeletConfigReserved,
+) (string, error) {
```
put this in multi lines as well

MichaelEischer replied:
Is changed.

rfranzke commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig_test.go:
```diff
@@ -1252,7 +1253,7 @@ var _ = Describe("OperatingSystemConfig", func() {
 			p                   *gardencorev1beta1.Worker
 			nodeLocalDNSEnabled bool
-			kubeReserved        *gardencorev1beta1.KubeletConfigReserved
+			kubletConfig        *gardencorev1beta1.KubeletConfig
```
typo

rfranzke commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -1033,6 +1033,9 @@ func KeyV2(
 		if kubeReserved := kubeletConfiguration.KubeReserved; kubeReserved != nil {
 			data = append(data, fmt.Sprintf("%s-%s-%s-%s", kubeReserved.CPU, kubeReserved.Memory, kubeReserved.PID, kubeReserved.EphemeralStorage))
 		}
+		if eviction := kubeletConfiguration.EvictionHard; eviction != nil && eviction.MemoryAvailable != nil {
```
pls update https://github.com/gardener/gardener/blob/master/docs/usage/shoot_updates.md#rolling-update-triggers

MichaelEischer replied:
I've updated the documentation and also included the other `EvictionHard` fields.

rfranzke reviewed (changes requested):
> For evictionHard, I've so far only included memory, although we could also add the FS-based limits.

So, why not add it?

> I think that the changed hash calculation and switching between the versions should be E2E tested. However, I'm unsure how to proceed here. Create a new E2E test? Extend an existing one? In particular, the feature gate would have to be enabled, which probably isn't the case right now for the E2E tests?

You can enable the feature gate locally and in e2e tests already (`provider-local` should be already ready for this change, right?).

Otherwise, I would create a new e2e test (i.e., creating a dedicated shoot with `v1`, and then see it being migrated to `v2` after a relevant change)?

MichaelEischer committed:
address review comments

MichaelEischer committed:
test and simplify CalcluateEffectiveKubeletConfiguation

MichaelEischer committed:
operatingsystemconfig: include all fields of EvictionHard in KeyV2

MichaelEischer committed:
document new node roll triggers

MichaelEischer committed:
add nodeAgentSecretName to worker.md docs

MichaelEischer committed:
enable feature gate WorkerPoolHashWithNodeAgentSecret in local setup

ScheererJ commented on pkg/apis/core/v1beta1/helper/helper.go:
```diff
@@ -1162,6 +1162,19 @@ func CalculateEffectiveKubernetesVersion(controlPlaneVersion *semver.Version, wo
 
+// CalcluateEffectiveResourceReservations if a shoot has a kubelet configuration specified at the worker group, return this,
+// otherwise the shoot resource reservations
+func CalcluateEffectiveResourceReservations(shootKubelet *gardencorev1beta1.KubeletConfig, workerKubernetes *gardencorev1beta1.WorkerKubernetes) *gardencorev1beta1.KubeletConfigReserved {
```
```suggestion
// CalculateEffectiveResourceReservations returns the worker group specific kubelet configuration (if available),
// otherwise the shoot resource reservations are returned
func CalculateEffectiveResourceReservations(shootKubelet *gardencorev1beta1.KubeletConfig, workerKubernetes *gardencorev1beta1.WorkerKubernetes) *gardencorev1beta1.KubeletConfigReserved {
```

MichaelEischer replied:
I've replaced the function with `CalcluateEffectiveKubeletConfiguation` in a later commit, but didn't backport the change to the earlier commits.

ScheererJ replied:
1. The typo in the function name is still present in the latest commit.
2. The description comment still does not read like a sentence in the latest commit.

Please adapt both.

MichaelEischer replied:
Oh, you're right, sorry. `CalculateEffectiveKubeletConfiguration` even had a second typo in its name :see_no_evil: .

ScheererJ commented on pkg/apis/core/v1beta1/helper/helper.go:
```diff
@@ -1162,6 +1162,19 @@ func CalculateEffectiveKubernetesVersion(controlPlaneVersion *semver.Version, wo
+			return reserved.DeepCopy()
+		}
+	} else if shootKubelet != nil && shootKubelet.KubeReserved != nil {
+		return shootKubelet.KubeReserved.DeepCopy()
```
Any particular reason for the `DeepCopy()` calls or just general defensiveness? It does not seem to be necessary as the values should only be used for calculations. As far as I have seen, the gardener code base usually omits unnecessary `DeepCopy()` calls.

@rfranzke Feel free to chime in if I misunderstood this.

MichaelEischer replied:
The `DeepCopy()` call has also been replaced with `CalcluateEffectiveKubeletConfiguation()`

ScheererJ replied:
`DeepCopy()` calls have been removed in later commits.

ScheererJ commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -60,7 +60,7 @@ const (
 
 // LatestHashVersion is the latest version support for calculateKeyVersion. Exposed for testing.
-var LatestHashVersion = 1
+var LatestHashVersion = 2
```
More of a meta question: Can this pull request come with the same release as https://github.com/gardener/gardener/pull/9846 or is it supposed to be following in at least one release after https://github.com/gardener/gardener/pull/9846? In other words, should it work to go from completely old calculation without `pool-hash` secret to this variant with version 2?

MichaelEischer replied:
I've added code in https://github.com/gardener/gardener/blob/615c9bab58957ff8fa39929ab43f6c8de6dc1c39/cmd/gardenlet/app/migration.go#L119 to create a migration secret on startup of a new gardenlet version. That migration secret always uses hash version 1. Thus, it should be safe to rollout both changes at once as old shoots simply stay at the old hash version 1.

ScheererJ commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -933,17 +941,20 @@ func (o *operatingSystemConfig) calculateKeyForVersion(version int, worker *gard
 var CalculateKeyForVersion = calculateKeyForVersion
 
-func calculateKeyForVersion(oscVersion int, kubernetesVersion *semver.Version, worker *gardencorev1beta1.Worker) (string, error) {
+func calculateKeyForVersion(oscVersion int, kubernetesVersion *semver.Version, credentialsRotation *gardencorev1beta1.ShootCredentialsRotation, worker *gardencorev1beta1.Worker, nodeLocalDNSEnabled bool, kubeReserved *gardencorev1beta1.KubeletConfigReserved) (string, error) {
```
Even though it was not introduced in this pull request, I propose to rename `oscVersion` to either just `version` or `oscKeyVersion`.
From my point of view `oscVersion` is misleading as it is in no way the operating system config version.

MichaelEischer replied:
I've renamed it to `version`.

ScheererJ commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -968,6 +979,57 @@ func KeyV1(workerPoolName string, kubernetesVersion *semver.Version, criConfig *
 }
 
+// KeyV2 returns the key that can be used as secret name based on the provided worker name, Kubernetes version and CRI
+// configuration.
```
Please adapt the documentation to the reality. In contrast to `KeyV1` there are more parameters that play a role in `KeyV2`.

MichaelEischer replied:
I've roughly included all relevant fields in the description, but didn't go into details regarding the `kubelet` fields.

Fixing the description, I've noticed that the I've lost the worker pool name in the hash, which could cause conflicts between OSC secrets from different worker pools. I've added the name in a new separate commit.

ScheererJ commented on pkg/gardenlet/operation/botanist/operatingsystemconfig.go:
```diff
@@ -110,6 +110,13 @@ func (b *Botanist) DeployOperatingSystemConfig(ctx context.Context) error {
+		credRotStatus = shoot.Status.Credentials.Rotation
+	}
+	b.Shoot.Components.Extensions.OperatingSystemConfig.SetCredentialsRotationStatus(credRotStatus)
+
```
Why not make it more concise? The `nil` check for shoot is not necessary. It is already assumed to be non-nil prior to this code, e.g. in `DefaultOperatingSystemConfig` or directly after this in `ShootEnablesSSHAccess(...)`. Calling `SetCredentialRotationStatus` with `nil` as parameter does not seem to be required either.

```suggestion
	shoot := b.Shoot.GetInfo()
	if shoot.Status.Credentials != nil {
		b.Shoot.Components.Extensions.OperatingSystemConfig.SetCredentialsRotationStatus(shoot.Status.Credentials.Rotation)
	}
	

```

MichaelEischer replied:
Oh, indeed. Thanks!

ScheererJ commented on docs/deployment/feature_gates.md:
```diff
@@ -36,6 +36,7 @@ The following tables are a summary of the feature gates that you can set on diff
 | ShootManagedIssuer                 | `false` | `Alpha` | `1.93` |        |
 | VPAForETCD                         | `false` | `Alpha` | `1.94` |        |
 | VPAAndHPAForAPIServer              | `false` | `Alpha` | `1.95` |        |
+| WorkerPoolHashWithNodeAgentSecret  | `false` | `Alpha` | `1.96` |        |
```
I find the name of the feature flag hard to understand. Why not call it `NewWorkerPoolHash` or something along those lines? In particular, I would leave the node agent secret out of the feature flag name. After all, the new worker hash is the important part.

MichaelEischer replied:
I've renamed the feature flag.

ScheererJ commented on docs/deployment/feature_gates.md:
```diff
@@ -208,3 +209,4 @@ A *General Availability* (GA) feature is also referred to as a *stable* feature.
 | ShootManagedIssuer              | `gardenlet`                       | Enables the shoot managed issuer functionality described in GEP 24.                                                                                                                                                                                                                                                                     ...
 | VPAForETCD                      | `gardenlet`, `gardener-operator`  | Enables VPA for `etcd-main` and `etcd-events`, regardless of HVPA enablement.                                                                                                                                                                                                                                                           ...
 | VPAAndHPAForAPIServer           | `gardenlet`, `gardener-operator`  | Enables an autoscaling mechanism for `kube-apiserver` of shoot or virtual garden clusters, and the `gardener-apiserver`. They are scaled simultaneously by VPA and HPA on the same metric (CPU and memory usage). The pod-trashing cycle between VPA and HPA scaling on the same metric is avoided by configuring the HPA to scale on a ...
+| WorkerPoolHashWithNodeAgentSecret | `gardenlet`                       | Enables usage of the new worker pool hash calculation. All provider extension must be upgraded to support this feature first.enablement.                                                                                                                                                                                              ...
```
Should we not include more functional details why this might be interesting/what this means?
The most notable change will be that changes to `kubeReserved` will result in a node rollout. Apart from that, it should not have notable changes as far as I understand.

MichaelEischer replied:
I've added a remark regarding kubeResered, evictionHard and cpuManagerPolicy

ScheererJ commented on pkg/features/features.go:
```diff
@@ -84,7 +84,13 @@ const (
+	VPAAndHPAForAPIServer featuregate.Feature = "VPAAndHPAForAPIServer"
+
+	// WorkerPoolHashWithNodeAgentSecret enables a new calculation method for the worker pool hash used by provider extensions. All
+	// such extensions must be upgraded to this Gardener version before enabling the feature gate.
```
`this Gardener version` is not well defined. Please adapt maybe similarly to how it is worded in `feature_gates.md`.

MichaelEischer replied:
The description is now synced with that in `feature_gates.md`.

ScheererJ commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -209,6 +209,8 @@ type Data struct {
 	// Content is the actual cloud-config user data.
 	// TODO(rfranzke): Remove this Content field after v1.100 is released.
 	Content string
+	// IncludeHashInWorker state whether the WorkerPool definition must include the GardenerNodeAgentSecretName
```
```suggestion
	// IncludeHashInWorker states whether the WorkerPool definition must include the GardenerNodeAgentSecretName
```

Apart from this, I do not understand the description. If we just use the secret name as hash, which in itself is a bit misleading, this should be stated explicitly.
I had expected that this attribute tells me whether a hash should be included in the worker pool or not. However, a secret name is not necessarily a hash. Therefore, I am confused.

MichaelEischer replied:
I've renamed the variable to `IncludeSecretNameInWorkerPool` and slightly changed the description. Is that enough or should I describe the context a bit further?

ScheererJ commented on pkg/component/extensions/worker/worker.go:
```diff
@@ -225,6 +225,11 @@ func (w *worker) deploy(ctx context.Context, operation string) (extensionsv1alph
 
+		var oscHash *string
+		if oscConfig.Init.IncludeHashInWorker {
+			oscHash = &oscConfig.Init.GardenerNodeAgentSecretName
```
I would have expected a short comment here stating why a secret name can be used as hash in this context.

MichaelEischer replied:
The variable is now called `nodeAgentSecretName`.

ScheererJ commented on pkg/apis/extensions/v1alpha1/types_worker.go:
```diff
@@ -125,10 +125,10 @@ type WorkerPool struct {
 	// Name is the name of this worker pool.
 	Name string `json:"name"`
-	// OSCHash is a hash of selected aspects of the OperatingSystemConfig. If the hash changes, then the
+	// NodeAgentSecretName is a hash of selected aspects of the OperatingSystemConfig. If the hash changes, then the
```
I find the description confusing. `NodeAgentSecretName is a hash` raises more questions than it answers. Documentation should clarify and not confuse. Please adapt.

Maybe you can state something along the lines of "NodeAgentSecretName is uniquely identifying selected aspects of the OperatingSystemConfig. If it changes, ...".

MichaelEischer replied:
Changed. Thanks.

ScheererJ commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -1048,8 +1049,12 @@ func KeyV2(
+			for _, value := range []*string{eviction.ImageFSAvailable, eviction.ImageFSInodesFree, eviction.MemoryAvailable, eviction.NodeFSAvailable, eviction.NodeFSInodesFree} {
+				parts = append(parts, ptr.Deref(value, ""))
+			}
+			data = append(data, strings.Join(parts, "-"))
```
Nit: I would have preferred a single `Sprintf(...)` instead of this boiler plate. Especially, the magic numbering in the sizing of the `parts` array might become useless over time if the amount of attributes changes.

MichaelEischer replied:
I've converted it into a sprintf statement. At least in the way the `eviction` fields are currently concatenated, adding a new field would result in a different hash. So the array size actually can't change.

ScheererJ commented on docs/usage/shoot_updates.md:
```diff
@@ -77,6 +77,9 @@ Apart from the above mentioned triggers, a rolling update of the shoot worker no
 * `.spec.kubernetes.version` (except for patch version changes)
+* `.spec.kubernetes.kubelet.kubeReserved` (unless a worker pool-specific value is set; if feature gate `WorkerPoolHashWithNodeAgentSecret` enabled)
+* `.spec.kubernetes.kubelet.evictionHard` (unless a worker pool-specific value is set; if feature gate `WorkerPoolHashWithNodeAgentSecret` enabled)
+* `.spec.kubernetes.kubelet.cpuManagerPolicy` (unless a worker pool-specific value is set; if feature gate `WorkerPoolHashWithNodeAgentSecret` enabled)
```
Do you mind changing the order to `if ...; unless ...`? I find it easier to comprehend this way.

MichaelEischer replied:
Reordered.

ScheererJ reviewed (commented):
Thanks a lot for this enhancement. I have a few questions/remarks.

MichaelEischer committed:
address review comments

MichaelEischer committed:
operatingsystemconfig: add missing worker.name to keyV2

ScheererJ commented on pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go:
```diff
@@ -968,6 +1000,75 @@ func KeyV1(workerPoolName string, kubernetesVersion *semver.Version, criConfig *
+		result += utils.ComputeSHA256Hex([]byte(v))
+	}
+
+	return fmt.Sprintf("gardener-node-agent-%s", utils.ComputeSHA256Hex([]byte(result))[:16])
```
Why does the name no longer contain the worker pool name? In local tests I could prevent a shoot from reconciling when having two worker pools with the same settings leading to the same hash and hence the same secret name. This would not have happened if the worker pool name would be included here as well.

/hold as it might break clusters.

MichaelEischer replied:
removing the name was unintentional, see https://github.com/gardener/gardener/pull/9865#discussion_r1629802706

MichaelEischer replied:
I've fixed my fix from yesterday to just include the worker pool name as part of the secret as in KeyV1 (and as intended in https://github.com/gardener/gardener/issues/9699 ), arghh.

MichaelEischer replied:
I think I've found why the workerPool.Name was missing. The idea in https://github.com/gardener/gardener/issues/9699 was to remove the worker image name, which I've apparently mixed up with the workerImage.Name as both are called "local" in the local setup.

ScheererJ replied:
/hold cancel

MichaelEischer committed:
e2e: check osc hash version after gardener upgrade

MichaelEischer committed:
fix typos

MichaelEischer committed:
Revert "operatingsystemconfig: add missing worker.name to keyV2"

This reverts commit 121a4bc17c7fc648a2b319bebb6322a301375194.

MichaelEischer committed:
operatingsystemconfig: add missing worker.Name to KeyV2

MichaelEischer committed:
e2e: reconcile shoot before check osc hash secret content

MichaelEischer committed:
operatingsystemconfig: remove workerImageName for hash version 2

MichaelEischer committed:
e2e: fix upgrade migration check

timuthy commented on docs/usage/shoot_updates.md:
```diff
@@ -85,6 +88,9 @@ The complete list of fields that trigger a rolling update:
 * `.spec.provider.workers[].providerConfig`
```
How about to mention `providerConfig` is not included in v2?

For the sake of clarity, it might be worth creating two separate lists, one for v1 and one fore v2?

MichaelEischer replied:
I've extracted the `NewWorkerPoolHash` specific fields into a separate block.

If `NewWorkerPoolHash` is enabled and a shoot has been reconciled once, then changes to fields included only in v2 will also trigger a node roll. (The v1 and v2 hashes are by then included in the "worker-pools-operatingsystemconfig-hashes" secret, which allows detecting the v2 hash change, even though the worker pool name still uses v1). From the user perspective only the `providerConfig` will behave slightly different, as it is no longer included in v2 hashes but requires extensions to explicitly decide which fields of the providerConfig to include. So, from the perspective of a user it's mostly relevant whether the feature gate is enabled; whereas the hash version used by a worker pool is not really relevant (only for the providerConfig field).

timuthy commented on docs/deployment/feature_gates.md:
```diff
@@ -36,6 +36,7 @@ The following tables are a summary of the feature gates that you can set on diff
 | ShootManagedIssuer                 | `false` | `Alpha` | `1.93` |        |
 | VPAForETCD                         | `false` | `Alpha` | `1.94` |        |
 | VPAAndHPAForAPIServer              | `false` | `Alpha` | `1.95` |        |
+| NewWorkerPoolHash                  | `false` | `Alpha` | `1.97` |        |
```
```suggestion
| NewWorkerPoolHash                  | `false` | `Alpha` | `1.98` |        |
```

Hopefully last version change 😄

MichaelEischer replied:
Changed, along with a few other places.

timuthy commented on docs/deployment/feature_gates.md:
```diff
@@ -208,3 +209,4 @@ A *General Availability* (GA) feature is also referred to as a *stable* feature.
 | ShootManagedIssuer              | `gardenlet`                       | Enables the shoot managed issuer functionality described in GEP 24.                                                                                                                                                                                                                                                                     ...
 | VPAForETCD                      | `gardenlet`, `gardener-operator`  | Enables VPA for `etcd-main` and `etcd-events`, regardless of HVPA enablement.                                                                                                                                                                                                                                                           ...
 | VPAAndHPAForAPIServer           | `gardenlet`, `gardener-operator`  | Enables an autoscaling mechanism for `kube-apiserver` of shoot or virtual garden clusters, and the `gardener-apiserver`. They are scaled simultaneously by VPA and HPA on the same metric (CPU and memory usage). The pod-trashing cycle between VPA and HPA scaling on the same metric is avoided by configuring the HPA to scale on a ...
+| NewWorkerPoolHash               | `gardenlet`                       | Enables usage of the new worker pool hash calculation. The new calculation supports rolling worker pools if `kubeReserved`, `evicitonHard` or `cpuManagerPolicy` in the `kubelet` configuration are changed. All provider extensions must be upgraded to support this feature first.                                                    ...
```
```suggestion
| NewWorkerPoolHash               | `gardenlet`                       | Enables usage of the new worker pool hash calculation. The new calculation supports rolling worker pools if `kubeReserved`, `evicitonHard` or `cpuManagerPolicy` in the `kubelet` configuration are changed. All provider extensions must be upgraded to support this feature first. Existing worker pools are not immediately migrated to the new hash variant, since this would trigger a replacements of all nodes. The migration happens, when a rolling update was triggered according to the old hash version calculation.                                                                                                                                                                                                                                                                                  |
```

Might be an important information for operator. They can safely switch on the feature gate.

MichaelEischer replied:
Good idea, thanks. I've slightly tweaked the wording of the last sentence. If `NewWorkerPoolHash` is enabled, then after the shoot has been reconciled once, then a node pool roll can also be triggered by the new fields.

timuthy reviewed (commented):
Very nice contribution, thank you 👏 

The changes look good to me, I could also test them successfully.

MichaelEischer committed:
update version in comments

MichaelEischer committed:
improve NewWorkerPoolHash documentation

MichaelEischer committed:
drop e2e check for worker pool hash migration secret

Gardener 1.97 already contains the code to create the worker pool hash
secret. Thus, no migration secret is created any more, which makes the
check obsolete.

timuthy reviewed (commented):
Thanks for the contribution!
/lgtm
/approve
